### ISSUE TITLE
Allow default format for timestamps

### ISSUE THREAD

**Issue by truthdoug** [dbcli/pgcli#1402]
## Description
It'd be nice if I could set a default format for timestamp columns. For example, rather than seeing `2023-04-27 15:16:38.125568+00`, I'd prefer to see `2023-04-27 15:16:38` (or maybe even without the seconds).

**Comment by j-bennet** [comment]
Hi @truthdoug, this might be a nice feature. I can't promise you when and if this would get implemented though, since everyone working on `pgcli` is contributing in their own free time. However, if you're interested in implementing the feature yourself, take a look at preprocessors in `cli_helpers`:

https://github.com/dbcli/cli_helpers/blob/main/cli_helpers/tabular_output/preprocessors.py

You would need to add a new preprocessor that would do the formating for date fields, and then reference that processor in whatever adapter you want it, certainly https://github.com/dbcli/cli_helpers/blob/main/cli_helpers/tabular_output/tabulate_adapter.py, but maybe the others too, like the delimited output and tsv.

We are happy to accept contributions, and it might get you the feature quicker.

**Comment by dbaty** [comment]
I would find it _very_ unexpected to see two distinct values (e.g. "2023-01-01 00:00:00.**12345**" and "2023-01-01 00:00:00.**67890**") both appear the same, as "2023-01-01 00:00:00". I guess that there may be some use cases where the timestamps are such that having millisecond- or second-scale precision is irrelevant, but that seems very specific.

Also, PostgreSQL function `date_trunc` already lets us print a date with any precision:

```sql
> select now(), date_trunc('second', now())
+-------------------------------+------------------------+
| now                           | date_trunc             |
|-------------------------------+------------------------|
| 2024-03-05 09:41:17.807009+00 | 2024-03-05 09:41:17+00 |
+-------------------------------+------------------------+
```

I am not sure that we really want to introduce a new format in pgcli for such rare use cases.

**Comment by truthdoug** [comment]
Yeah, I agree that would be jarring.... but that's why it'd be configurable. For the data I'm looking at, the columns are general significant to the day and possibly the minute. The seconds and microseconds never matter.

I'm aware of `date_trunc` and other PG functions that allow me to reformat dates. I proposed this as a feature that I thought might be compatible with how `pgcli` helps with better visuals.

I may take a stab of this if I find some of that "free time" I hear other people talking about. 

I think my dream goal would be to have default formatting by column name so that I would get an expected format on any query. Something like:

```
[date formats]
birthdate = 'YYYY-MM-DD'
registration = 'YYYY-MM-DD HH24:MI'
graduation_year = 'YYYY'
```

**Comment by j-bennet** [comment]
> I may take a stab of this if I find some of that "free time" I hear other people talking about.

I see you have kids, @truthdoug. 😂

**Comment by j-bennet** [comment]
@dbaty There's no harm in making this feature configurable. When it's configured to something non-default though, it would be good to have a visual indication of it (for example in the status bar). I can imagine the user configuring it to truncate timestamps to date, forgetting about it, and then being alarmed when all timestamps look the same.

**Comment by dougharris** [comment]
Sometimes I find free time between Xmas and New Years and pick an itch to scratch. This year, it's this issue.

I've submitted PRs for both dbcli and cli_helpers (see links automatically added above). With this change, one can add per-column lines to a new section of pgcli's config file to specify format strings.

For example, if you add this to `~/.config/pgcli/config` (or wherever your pgcli config file is):
```ini
[column_date_formats]
created_date = "%Y-%m-%d"
last_visit =  "%I:%M:%S %p %m/%d/%y" # I'm not sure _why_ anybody would choose this format though
```

and then query `select username, created_date, last_visit from users where username = 'foo';` your results would look something like
```
+----------+--------------+----------------------+
| username | created_date | last_visit           |
|----------+--------------+----------------------|
| foo      | 2019-11-19   | 05:00:00 AM 12/17/19 |
+----------+--------------+----------------------+
```

In my use cases, there are some columns where I really only ever need to know granularity to the day, other times I need more.

This is completely optional. If you never edit your pgcli config file and it doesn't have the `column_date_formats`, things will work as they always have. If you select a timestamp column and there's no line for it in `column_date_formats`, it will show with the default formatting.

What if you need to show the full data but don't want to exit pgcli, edit your config, and restart? The preprocessor is based on the column name as returned from the query, so using a column alias would ignore any configuration formatting you've provided. If I needed to know full details on the created_date column, I could alter my query to `select username, created_date as raw_created, last_visit from users where username = 'foo';`

**Comment by dbaty** [comment]
A `column_date_formats` has been added in 72589311129f4e73e1d71b2f15733ff448e2dfb8, which is included in pgcli since version 4.3.0.
